### DISCUSSION TITLE
How to enable the always-on analysis function in settings

### DISCUSSION THREAD

**Discussion by romanholidaypancakes** [Vector35/binaryninja-api#5859] [Q&A] [answered]
`this function is generating excessive incremental updates ...`

![image](https://github.com/user-attachments/assets/4fe71e32-02da-4ac1-b9e0-1ef8a1d9529e)

I don't want to click manually every time. I want it to be analyzed automatically.

**Comment by psifertex** [comment, marked as answer]
This is how you get infinite analysis loops that can't be stopped. I'm not sure this is something we will ever add as a setting but I will discuss it with others this week. 

That said, there is an API on the function you can use so if you REALLY want to, you can write a snippet to trigger the override for all functions. 

https://api.binary.ninja/binaryninja.function-module.html#binaryninja.function.Function.analysis_skipped

EDIT: You can actually get the behavior requested by setting `analysis.limits.maxFunctionUpdateCount = 0` in settings. However, this will result in infinite analysis that never ends and therefore isn't a good option.

**Reply by romanholidaypancakes** [reply]
I tried changing maxFunctionUpdateCount to 100000 but it didn't work, 
```
[Function] Skipping analysis for function at 0x14017e4d3 : Exceeds 'analysis.limits.maxFunctionUpdateCount'. If possible, please report this issue.
```
in fact there were many functions that contained valid code snippets (the files were obfuscated) after I clicked Force Update.
I'm currently trying the free version so I can't execute the script.

**Reply by psifertex** [reply]
Can you share the file in question?

**Reply by romanholidaypancakes** [reply]
[demo.zip](https://github.com/user-attachments/files/16661877/demo.zip)
password: 111

**Reply by psifertex** [reply]
Thanks for the file. It's as I suspected--you can actually get exactly what you want by setting that value to `0` but what you'll have happen instead is basically analysis never stops. There's some pathalogical case that's causing multiple functions to trigger updates in the other and it's never able to reach a steady state. This is exactly why this setting exists in the first place. It's basically impossible to guarantee that this situation can't occur based on our current analysis design and so that's why we have this particular setting.

**Reply by romanholidaypancakes** [reply]
There are still many functions to click :
![image](https://github.com/user-attachments/assets/4b9902ce-5511-45b7-9848-f6772672c0ea)

My analysis settings :
![image](https://github.com/user-attachments/assets/0edaecd5-48a6-44b6-9b4c-8d877e472f68)

Binary Ninja Version: 4.1.5902 free, f2165c5d

**Reply by plafosse** [reply]
I think you need to increase your `maxFunctionAnalysisTime` or set it to `0` to let it take as much time as it wants.

**Reply by xusheng6** [reply]
I looked at the binary and I think what you want is actually NOT to force analyze those functions. The binary is probably packed and those bytes just happen to be able to be disassembled. You should at least unpack the binary first before you continue to statically analyze it.

That said, we can still discuss whether we indeed want a way to let the user somehow force the analysis

**Reply by romanholidaypancakes** [reply]
> I looked at the binary and I think what you want is actually NOT to force analyze those functions. The binary is probably packed and those bytes just happen to be able to be disassembled. You should at least unpack the binary first before you continue to statically analyze it.
> 
> That said, we can still discuss whether we indeed want a way to let the user somehow force the analysis

From my experience this file is not packed, it is just obfuscated and partially virtualized

**Reply by romanholidaypancakes** [reply]
> I think you need to increase your `maxFunctionAnalysisTime` or set it to `0` to let it take as much time as it wants.

The above picture example is the result after the analysis is completed: there are still functions that need to be forced to click (the file I tested is the file I uploaded)

**Reply by psifertex** [reply]
Did you change maxFunctionAnalysisTime to 0 as well? The last image shows that the individual functions are timing out, not that total overall analysis is timing out.

**Reply by romanholidaypancakes** [reply]
> Did you change maxFunctionAnalysisTime to 0 as well? The last image shows that the individual functions are timing out, not that total overall analysis is timing out.


Setting it to 0 is useful. However, the following problems still exist:

1. When I analyze a database that I have saved before, even if I trigger re-analysis in the menu bar, the functions that needed to be clicked before still need to be clicked afterwards
2. I encountered the problem you reminded me of, and the file was still being analyzed after opening it for several hours (it seems to be stuck in an endless loop, but the analysis results are much better than before)

**Reply by psifertex** [reply]
Unfortunately the only work-around for an existing database to mass apply changes to existing functions would be to use the API, we're not considering further changes that would impact free.
